// expo-cli/build/commands/apply/configureProjectAsync.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = configureManagedProjectAsync;
exports.expoManagedPlugins = void 0;

function _config() {
  const data = require("@expo/config");

  _config = function () {
    return data;
  };

  return data;
}

function _configPlugins() {
  const data = require("@expo/config-plugins");

  _configPlugins = function () {
    return data;
  };

  return data;
}

function _xdl() {
  const data = require("@expo/xdl");

  _xdl = function () {
    return data;
  };

  return data;
}

function _log() {
  const data = _interopRequireDefault(require("../../log"));

  _log = function () {
    return data;
  };

  return data;
}

function _ConfigValidation() {
  const data = require("../eject/ConfigValidation");

  _ConfigValidation = function () {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Expo managed packages that require extra update.
// These get applied automatically to create parity with expo build in eas build.
const expoManagedPlugins = ['expo-app-auth', 'expo-av', 'expo-background-fetch', 'expo-barcode-scanner', 'expo-brightness', 'expo-calendar', 'expo-camera', 'expo-contacts', 'expo-image-picker', 'expo-file-system', 'expo-location', 'expo-media-library', // 'expo-notifications',
'expo-screen-orientation', 'expo-sensors', 'expo-task-manager', 'expo-local-authentication']; // Plugins that need to be automatically applied, but also get applied by expo-cli if the versioned plugin isn't available.
// These are split up because the user doesn't need to be prompted to setup these packages.

exports.expoManagedPlugins = expoManagedPlugins;
const expoManagedVersionedPlugins = [// 'expo-splash-screen',
// 'expo-facebook',
// 'expo-branch',
// 'expo-updates',
// 'expo-ads-admob',
'expo-apple-authentication', 'expo-document-picker', 'expo-firebase-analytics', 'expo-firebase-core', 'expo-google-sign-in' // 'expo-dev-menu',
// 'expo-dev-launcher',
];

const withOptionalPlugins = (config, plugins) => {
  return plugins.reduce((prev, plugin) => {
    return (0, _configPlugins().withStaticPlugin)(prev, {
      plugin,
      // If a plugin doesn't exist, do nothing.
      fallback: config => config
    });
  }, config);
};

function withManagedPlugins(config) {
  return withOptionalPlugins(config, [...new Set(expoManagedVersionedPlugins.concat(expoManagedPlugins))]);
}

async function configureManagedProjectAsync({
  projectRoot,
  platforms
}) {
  // let config: ExpoConfig;
  let {
    exp: config
  } = (0, _config().getConfig)(projectRoot, {
    skipSDKVersionRequirement: true,
    isModdedConfig: true
  });
  const expoUsername = process.env.EAS_BUILD_USERNAME || (await _xdl().UserManager.getCurrentUsernameAsync()); // Add all built-in plugins first because they should take
  // priority over the unversioned plugins.

  config = (0, _configPlugins().withExpoVersionedSDKPlugins)(config, {
    expoUsername
  });
  config = withManagedPlugins(config);

  if (platforms.includes('ios')) {
    // Check bundle ID before reading the config because it may mutate the config if the user is prompted to define it.
    const bundleIdentifier = await (0, _ConfigValidation().getOrPromptForBundleIdentifier)(projectRoot);
    config.ios.bundleIdentifier = bundleIdentifier; // Add all built-in plugins

    config = (0, _configPlugins().withExpoIOSPlugins)(config, {
      bundleIdentifier
    });
  }

  if (platforms.includes('android')) {
    // Check package before reading the config because it may mutate the config if the user is prompted to define it.
    const packageName = await (0, _ConfigValidation().getOrPromptForPackage)(projectRoot);
    config.android.package = packageName; // Add all built-in plugins

    config = (0, _configPlugins().withExpoAndroidPlugins)(config, {
      package: packageName
    });
  } // compile all plugins and mods


  config = await (0, _configPlugins().compileModsAsync)(config, {
    projectRoot,
    platforms
  });

  if (_log().default.isDebug) {
    _log().default.debug();

    _log().default.debug('Evaluated config:'); // @ts-ignore: mods not on config type


    const {
      mods,
      ...rest
    } = config;

    _log().default.info(JSON.stringify(rest, null, 2));

    _log().default.info(mods);

    _log().default.debug();
  }

  return config;
}
